### ISSUE TITLE
Analyzer suggestion: Use lambda expression

### ISSUE THREAD

**Issue by Neme12** [dotnet/roslyn#24786]
In the spirit of analyzers that suggest the use of newer language features, I propose an analyzer that would flag anonymous methods and offer to replace them with lambda expressions. Lambda expressions are definitely not a new feature today but they once were and my guess is that if they were introduced after Roslyn, there would have been an analyzer for this already. So maybe it's not a bad idea to revisit established language features that in some or even all cases replace even older ones.

This analyzer would be very similar to "Use local function" and would also have a code style option to "Prefer lambda expressions to anonymous methods", which would be set as a suggestion by default.

Should this be a 2-way code fix?
- On the one hand, this is a very local & *mostly* syntactic change similar to "use expression body", which would suggest that there should also be a code fix to convert it back. That's one way of looking at it.
- On the other hand, is this really a code style preference? If my preference is to use expression-bodied members, it's very reasonable that I might want to change it back into a block in the future if an expression won't cut it anymore. But are there any reasons you might want to go back to an anonymous method? Are there any coding styles that tell you to prefer anonymous methods over lambdas? And do we really want people to use them? Maybe this is just not worth it. Also, aren't lambda expressions in this case just the "new language feature" that we suggest as a better way of doing things? In that sense, there shouldn't be a code fix to go back. We don't offer one to undo pattern matching for example.

Other questions:
- would this try to remove parameter types when possible?
- would this also try to convert to a simple lambda when possible?

**Comment by CyrusNajmabadi** [comment]
> Should this be a 2-way code fix?

No.  Anonymous-methods are essentially a mistake of the language.  Something i know i regret when i worked on creating them for C# 2.0.  They were a rush to solve the problem before we made the much better solution of 'lambdas' for C# 3.0.

The IDE should not ever be working to move code toward them.

**Comment by CyrusNajmabadi** [comment]
> Would this try to remove parameter types when possible?

By default yes.  That's the idiomatic way that we've shown with lambdas since the beginning.  They're intended to be lightweight, and the default change should produce the lightest lambdas possible.

> Would this also simplify to a simple lambda when possible?

Yes.  Same reason as above.

> Possibly even convert the block to an expression, if possible? 

Yes.  Same reason as above.

> Should lambda parameter types be faded out when they're not necessary with an option to remove them? Should parentheses around a single lambda parameter also be marked as unnecessary?

Potentially.  But this would be another codestyle/feature.

**Comment by Neme12** [comment]
@CyrusNajmabadi Do you think it would be a good addition then?

**Comment by CyrusNajmabadi** [comment]
> Do you think it would be a good addition then?

WRT having the feature that converts anonymous methods to lambdas.  I think the idea is nice... but very minor.  We've had lambdas since 3.0.  And we only had anonymous methods in 2.0.  Their usage is very limited,  so very few people would benefit from a tool to fix them up.   So, for example, i would not personally spend time on it since i don't think the benefit would be large enough.  I leave it to @dotnet/roslyn-ide  and @kuhlenh to decide if they would take such a feature if it was written.

WRT: "Should lambda parameter types be faded out when they're not necessary with an option to remove them? Should parentheses around a single lambda parameter also be marked as unnecessary?"  I think this would be *very* nice.  Lambdas are in widespread use, and i can certainly see someone wanting to enforce a certain style here and have the tooling do the work for them.  i.e. someone who always wants parenthesized lambdas with types, vs. someone who prefers non-parenthesizes/non-types as much as possible.  This is similar to the 'var' style and would fit very well into the set of styles/features that roslyn has.  This one def seems worth doing, and i think the team would be more likely to take this.  Again for @dotnet/roslyn-ide and @kuhlenh to decide.

**Comment by CyrusNajmabadi** [comment]
in other words: help customers move their **large** code investment forward when newer and better language features come out.  And help them codify the types of common stylistic preferences we see with widespread use out there.
